Implement Swift backend

ScrelliCopter opened the pull request:
Add a syntax file for the Swift programming language, as well as examples, tests, and documentation.

<details>

<summary>Outdated notes</summary>

Hand full of notes so far:
 - I've had to set character literals to `hex`, 'c' style character literals don't exist in Swift and the only way (that I'm aware of) to represent single-byte character literally is to construct it like `UInt8(ascii: "c")` (where "c" is a UnicodeScalar). Raw hex is not the end of the world and if we really wanted "native" character literals, support would have to explicitly be added to the C++ side and it probably isn't worth it. 
 - Case-ranges are incompatible with UnicodeScalar, and when I get around to tackling widechar they may not be viable to support over fixed width ints.
 - ~~Swift integer conversions are checked and will panic on out-of-range inputs, I've worked around this in the test by constructing the target type with `bitPattern: ` but this will only be valid for integer types of the same width so this clearly doesn't belong in the syntax file. I'll probably end up putting the onus on the user to deal with this but I'm not sure how to handle this in the conv test right now.~~
 - I've picked 2 space indentation for Swift code as it's the style used in the Swift compiler and recommended by a lot of style guides, the other lang backends don't seem to be consistent about indentation (some use tab instead of 4 spaces) so I went with my preference. If this is unjustifiable I can change the formatting to something else (tabs would be my preference in this case). All re2c files and scripts match the project style.

</details>

Related issue: #506

skvadrik commented:
Great work!

Using hex chars is fine.

Case-ranges feature doesn't have to be supported, although it's good to have. It may be suppoted only under certain conditions, like we do for OCaml (the language has different rules for chars and integers).

Regarding conversions, why would they be out of ranges? As I understand, conversion is needed because `YYPEEK` reads a character but `yych` has to be an integer to match against hex literals. This conversion should be really fast. But speed consideration aside, why would there be out of range and panic?

2-space formatting is fine, users can change it if they want.

ScrelliCopter commented:
> Case-ranges feature doesn't have to be supported, although it's good to have. It may be supported only under certain conditions, like we do for OCaml (the language has different rules for chars and integers).

For now the support works well, if UnicodeSingle or Character ends up being much more interesting to support it can default to off but stay an optional feature as it's handy for 1-byte use cases.

> Regarding conversions, why would they be out of ranges? As I understand, conversion is needed because `YYPEEK` reads a character but `yych` has to be an integer to match against hex literals. This conversion should be really fast. But speed consideration aside, why would there be out of range and panic?

The more I think about it the less sense it makes to worry. Guess I figured because in the 8-bit case any >=0x80 unsigned input could cause an assert but this is impossible in the test, and in the real world can be solved by ensuring the input is valid ASCII or just not using unsigned > signed conversion in the first place. I'll just stick to the default checked conversion.

> 2-space formatting is fine, users can change it if they want.

Good to know.

ScrelliCopter commented:
@skvadrik Marking this as reviewable now that the syntax file and examples/tests are together. I'll be concentrating on getting the build scripts hooked up, then once all reviews are dealt with I'll squash and hopefully this PR will be in good shape.

Current notes:
 - Everything was tested on both MacOS and Linux systems.
 - I'm strongly considering simplifying the header sample to just include `state.swift` as a source file. The whole module song and dance just to be able to write `import lexer` complicates the test shell script more than I'd like.
 - Couldn't figure out how to emit `code:yyshiftmtag` in a test, so that codegen is untested.
 - Enum types are currently a hack, there are some cases where it needs to be a UInt instead of an Int but the syntax file doesn't what type to give the enum. I've done my best to guess based off context clues but please let me know if there is a better way of handling this than what I'm doing.
 - The codegen tests are still built as libraries with explicit `@main` entry points (this was a hack to make it easier to use my IDE), I'm considering making these top-level code just for consistency as well. A different heuristic would need to be used to skip the unsupported tests but that's already trivially solved by other backends.
 - I'm still considering evaluating if a Swift native string can be used as yyinput and make and example from it but I might leave it for another PR as it might not be feasible and I doubt it would require syntax changes.

That's all I can think of for now, please let me know if any changes or extra tests are required.

skvadrik reviewed (commented):
Great work! Thanks a lot for your thorough approach.

> I'm strongly considering simplifying the header sample to just include state.swift as a source file. The whole module song and dance just to be able to write import lexer complicates the test shell script more than I'd like.

If you can, by all means please do simplify it. Header example is always a bit of a pain with modern languages that have their own packaging systems and assume certain file layout. For re2c it only matters that the header directive works in some useful way.

> Couldn't figure out how to emit code:yyshiftmtag in a test, so that codegen is untested.

That's fine, this definition is a basic delegate to the user-defined YYSHIFTMTAG. If you follow my suggestion and get rid of function-like style, it will simplify even further.

> Enum types are currently a hack, there are some cases where it needs to be a UInt instead of an Int but the syntax file doesn't what type to give the enum. I've done my best to guess based off context clues but please let me know if there is a better way of handling this than what I'm doing.

See my comments below.

I won't go into more detailed review at the moment, let's first resolve the current issues. The rest seems good on a glance (and I don't know swift anyway!).

ScrelliCopter commented on include/syntax/swift:
```diff
@@ -0,0 +1,414 @@
+// supported feature lists -----------------------------------------------------
+
+supported_apis = ["simple", "generic", "record"];
+supported_api_styles = ["functions", "free-form"];
```
I figured since the language supports closures why not, but the Generic API exists so there's not really much of an advantage to having it. I wouldn't miss it.

ScrelliCopter replied:
Functions style is now removed, I also removed the test and added an unsupported test.

ScrelliCopter commented on include/syntax/swift:
```diff
@@ -0,0 +1,414 @@
+code:array_elem = array "[" index "]";
+
+//HACK: Conditions are assigned to `yystate` in loop/switch mode, so underlying type needs to be `UInt`.
+code:enum =
+    "enum " type ": " (.code_model.loop_switch & !.storable_state ? "UInt" : "Int") " {" nl indent
+        [elem: topindent "case " elem (.init ? " = " init) nl]
+    dedent "}" nl;
+
+code:enum_elem = type "." name ".rawValue";
```
My thought at the time was let constant pseudo-enums would have the same issue since they are typed and the root of the issue is that Swift prefers signed Int for array indices (unlike Rust which uses usize) which causes issues in certain cases because yystate is UInt. That is my recollection though I will double check my logic (in particular) that conditional tomorrow.

I'll also look into breaking this into let constants like you suggested, if that makes this less hacky, then why not.

skvadrik replied:
Ah, I see - in that case it's still better to have a constant pseudo-enum because you can move the condition to the default `YYCONDTYPE` definition and let the user override the type.

skvadrik replied:
... (as opposed to hardcoding it in the syntax file).

ScrelliCopter replied:
Alright, I moved the conditional to YYCONDTYPE and swapped to pseudo-enum let constants which solves the hard-coding issue. The conditional is still the same (loop_switch && !storable) and I tried rethinking the justification in the comment. I'll give it some more thought and see if a less potentially surprising behaviour can work here but this should fix most of the problem for now.

ScrelliCopter commented:
> If you can, by all means please do simplify it. Header example is always a bit of a pain with modern languages that have their own packaging systems and assume certain file layout. For re2c it only matters that the header directive works in some useful way.

Done. It was cathartic to delete all that shell 😄. I can either keep the @main entrypoint for this test (I added it back as swiftc handles multi-source invocations differently), or I can change the script to output `main.swift` instead of `example.swift` which will cause swiftc to treat the main file as the entry point (like it does for single-file invocations).
I'm leaning toward the former because while it's inconsistent it's less magical and won't shoot a user in the face who tries to `swiftc header.swift lexer/state.swift` manually. The only other alternative is to rename header.re to main.re but that just feels like swapping out one weird inconsistency for another.

skvadrik commented:
> > If you can, by all means please do simplify it. Header example is always a bit of a pain with modern languages that have their own packaging systems and assume certain file layout. For re2c it only matters that the header directive works in some useful way.
> 
> Done. It was cathartic to delete all that shell 😄. I can either keep the `@main` (edit: sorry for the ping) entrypoint for this test (I added it back as swiftc handles multi-source invocations differently), or I can change the script to output `main.swift` instead of `example.swift` which will cause swiftc to treat the main file as the entry point (like it does for single-file invocations). I'm leaning toward the former because while it's inconsistent it's less magical and won't shoot a user in the face who tries to `swiftc header.swift lexer/state.swift` manually. The only other alternative is to rename header.re to main.re but that just feels like swapping out one weird inconsistency for another.

Leaving `@main` for just the header test is totally fine, especially since it doesn't require any special handling due to `@main` in the run script. You can add a short comment right above `@main` that explains why it's needed for this particular test.

skvadrik commented:
Do you plan to update the Autotools build system as well?

skvadrik commented:
Ah, the failed build on CI tells me you forgot to add `bootstrap/src/default_syntax_swift.h`. You need to reconfigure with `-DRE2C_REBUILD_SYNTAX=1`.

ScrelliCopter commented:
> Ah, the failed build on CI tells me you forgot to add `bootstrap/src/default_syntax_swift.h`. You need to reconfigure with `-DRE2C_REBUILD_SYNTAX=1`.

Yep, I haven't committed any bootstrapped files just yet, I'll do that after Autotools is working on both my machines.

Which build system do you regenerate bootstrap files with? I've been doing it with CMake and there's a few non-swift (or the other recent PRs) differences with the files in the repo, I would like to keep the differences in build configuration as small as I can when generating files that get checked in.

skvadrik commented:
> Which build system do you regenerate bootstrap files with?

I usually work with Auto tools but there's no rule to do so, they should be on par with CMake. You can use build/__build.sh or build/__cmakebuild.sh and add configure options as needed to get the same setup as I use for development.

> I've been doing it with CMake and there's a few non-swift (or the other recent PRs) differences with the files in the repo, I would like to keep the differences in build configuration as small as I can when generating files that get checked in.

There may be some files that we forgot to regenerate for previous commits, or there may be just differences between bison / docutils versions.

If there are differences in lexers / parsers that are unrelated, you can regen them all in a separate commit before your main one.

For docutils (rst2man in particular), you need a slightly old version (0.19) as the new one has a bug: https://sourceforge.net/p/docutils/bugs/497. If that's a problem, just commit the doc changes without regenerating bootstrap files and I will regen them later with the old version.

ScrelliCopter commented:
Thanks for the information.

> For docutils (rst2man in particular), you need a slightly old version (0.19) as the new one has a bug: https://sourceforge.net/p/docutils/bugs/497. If that's a problem, just commit the doc changes without regenerating bootstrap files and I will regen them later with the old version.

This is also good to know, I'll check my versions and see if I need to downgrade.

ScrelliCopter commented:
@skvadrik Sorry that took so long, I did have to downgrade docutils and now the documentation looks like it's generating properly.
I went ahead and squashed all the Swift commits and rebased it on top of a clean build of master that had all the bootstrap directory regenerated, I matched my setup as close to yours as I could so presumably any differences are just other tool versions and also updates to re2c itself.
I regenerated with all the Swift changes and those are now in the Swift commit.

skvadrik commented on run_tests.py.in:
```diff
@@ -383,6 +385,7 @@ def run_test(test):
     switches = switches.replace('re2ocaml', 're2c --lang ocaml')
     switches = switches.replace('re2py', 're2c --lang python')
     switches = switches.replace('re2rust', 're2c --lang rust')
+    switches = switches.replace('re2swift, 're2c --lang swift')
```
Missing `'` after `'re2swift`.

ScrelliCopter replied:
There appears to be more problems than just that missing quote terminator, `tests` reports all the swift tests as missing and I'm not sure why yet.

skvadrik replied:
Ah, it's probably because you forgot to add .Swift file extension in the run_tests.py.in. I always get int this trap myself.

skvadrik replied:
(see line 302)

ScrelliCopter replied:
Okay, the missing issue was because I neglected to add the swift extensions to the glob. Now that that's fixed I'm still getting test failures, still not sure why yet.

ScrelliCopter replied:
Ohhhhh it's because diff is noticing the `sourceLocation`s are different because of how I generated the tests.

ScrelliCopter replied:
This particular issue should be solved now.

skvadrik commented:
> @skvadrik Sorry that took so long, I did have to downgrade docutils and now the documentation looks like it's generating properly.

Thanks!

> I went ahead and squashed all the Swift commits and rebased it on top of a clean build of master that had all the bootstrap directory regenerated, I matched my setup as close to yours as I could so presumably any differences are just other tool versions and also updates to re2c itself.

Bootstrap files look good now, thanks. There's a lot of trivial line changes but I get them as well if I touch the lexer and rebuild, which means someone else regenerated them with an in-tree build (I'm always using out-of-tree builds).

> I regenerated with all the Swift changes and those are now in the Swift commit.

Good.

skvadrik commented:
Let's get the tests pass on CI, and then I'll give the PR one last thorough review (although it looks good overall and I don't expect any serious obstacles).

ScrelliCopter commented:
> Let's get the tests pass on CI, and then I'll give the PR one last thorough review (although it looks good overall and I don't expect any serious obstacles).

Now that I've fixed all the issues I can, I can see that tests that produce warnings are causing the diffs to fail. Any ideas about this?

skvadrik commented:
> Now that I've fixed all the issues I can, I can see that tests that produce warnings are causing the diffs to fail. Any ideas about this?

You just need to append the warnings (and the generated header for the header example) at the end of the test output file. We do this for all languages, and the run scripts just grep away all the warnings before compiling/running the output files.

ScrelliCopter commented:
Alright, tests pass on both my machines now.

skvadrik commented:
There's a failing test on Windows due to incorrect file path separators in a `sourceLocation` directive. You can use `-i` option for this test to avoid this problem.

While at it, I noticed that you use `--api simple` option in the failed test. What's the point of this, simce simple API is on by default?

ScrelliCopter commented:
> There's a failing test on Windows due to incorrect file path separators in a `sourceLocation` directive. You can use `-i` option for this test to avoid this problem.

Done.

> While at it, I noticed that you use `--api simple` option in the failed test. What's the point of this, simce simple API is on by default?

I'm a bit confused, I made the syntax file default to the generic API, so the --api switch should be fine here right?

skvadrik commented:
> I'm a bit confused, I made the syntax file default to the generic API, so the --api switch should be fine here right?

Sorry for confusion, for some reason I assumes you made simple API the default. I think it should be the default for all new languages that support it. Historically a few languages have generic API by default, but that's because they were added before simple API.

skvadrik commented:
> When I put this together I mistakenly thought generic was the newer preferred API. In that case it makes the most sense to default to simple, I'll make it so.

Historically, for a long time there was only the C/C++ backend with what's currently called simple API. Then I added generic API as a way to get more flexibility out of re2c. Then I added Go and Rust backends which only supported generic API. And only recently I added syntax files and it became possible to define different API flavours for various backends wihout making re2c codegen a fragile mess of conditions and special cases. So, historically Go and Rust backends are stuck with generic API by default.

There're also Haskell and OCaml backends that cannot have simple API because they always need a struct (record) that is propagated between functions.

skvadrik commented on examples/swift/fill/01_fill.re:
```diff
@@ -0,0 +1,109 @@
+        re2c:api = record;
+        re2c:eof = 0;
+        re2c:variable:yyrecord = "self";
+        re2c:YYINPUT = "self.buffer";
```
Can you rename `buffer` field to `yyinput` and drop this config? Here in other tests.

ScrelliCopter replied:
Done for both fill tests.

skvadrik commented on examples/swift/headers/header.re:
```diff
@@ -0,0 +1,34 @@
+    re2c:tags          = 1;
+    re2c:yyfill:enable = 0;
+    re2c:yyrecord      = "self";
+    re2c:YYCURSOR      = "cur";
+    re2c:YYINPUT       = "str";
```
Can you rename these two as `yycursor` and `yyinput` and drop configurations? If you do so, please update the autogenerated header (it's easy to forget as the test only check the header pasted at the end of the output file).

ScrelliCopter replied:
Done.

skvadrik commented on include/syntax/swift:
```diff
@@ -0,0 +1,412 @@
+re2c:YYSTAGP = "yystagp";
+re2c:tags:expression = (.api.record ? yyrecord ".") sigil;
+re2c:tags:negative = (.api.generic ? "@@" : "-1");
+re2c:cond:divider = "/* *********************************** */";
```
I think you can make it empty string, this is used in goto/label mode which is unsupported for Swift backend.

ScrelliCopter replied:
Done.

skvadrik commented on include/syntax/swift:
```diff
@@ -0,0 +1,412 @@
+    [arg{-1}:   "_ " argname ": " argtype]
+    ")" (.type ? " -> " type) " {" nl
+        indent [stmt: stmt] dedent
+    "}" nl; // Can't specify throw
```
Did it cause any problems for you (not being able to specify throw)? It may totally happen that a new language requires stretching some syntax file configurations a bit, add new variables etc. Let me know if you find the current model too limiting.

ScrelliCopter replied:
I can't remember which test/example it was for, as I was able to figure out a different way of returning an error from the function.
Still, it would be nice if it was possible to specify throw as functions in Swift cannot throw unless marked throwing (and functions that don't shouldn't be), and it would be even nicer if it was possible to type the throw like [here](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/errorhandling/#Specifying-the-Error-Type) as constrained throws simplify error handling considerably, for example the user could create something like `enum LexerError: Error { case ... }` and only throw that from a recursive lexer, the compiler enforces this and callers don't have to worry about `any Error` forcing users to handle unknown errors or make the entire call stack throwing.

A hack around this would be to use the Result type although it has absolutely none of the syntax sugar that Rust has to make them convenient (Result in Swift exists for embedded Swift where small heap allocations are a non-starter).

It's not a requirement for this PR but if you're ever planning to extend the yyfn interface Swift would benefit from being able to specify throws greatly, perhaps an optional 3rd parameter could add it before `(.type ? " -> " type)`, though I'm not sure if this would provide value to other languages.

skvadrik replied:
I see, let's deal with it in after this PR is merged. I'll make a note for myself and mark this comment as resolved.

skvadrik replied:
https://github.com/skvadrik/re2c/commit/85d9384b2f95b3de0464b82a53680c8956dfc01c adds variable `attrs` and conditional `.attrs` to `code:fndecl` and `code:fndef`. Use the 3rd component of the first `YYFN` element to specify the attributes. Let me know if it works for Swift.

skvadrik replied:
(See the example in this commit for usage.)

ScrelliCopter replied:
Sweet, thanks a lot for this.

ScrelliCopter replied:
I'm testing out adding errors to the advanced recursive test, ran into an issue that I didn't think about when I proposed this. When throwing functions are called they *must* be preceded with the `try` keyword, you can convert an error to an Optional with `try?`, or inhibit error propagation with a panic using `try!`, but a throwing function call must be preceded with one of these three or the call will not compile.

Currently I've hacked this in with
```
code:fncall = topindent (.retval ? retval " = ") "try " name "(" [arg{0:-2}: arg ", "] [arg{-1}: arg] ")" nl;

code:tailcall = topindent (.retval ? "return ") "try " name "(" [arg{0:-2}: arg ", "] [arg{-1}: arg] ")" nl;
```
but this is only a bandaid solution and I'm not sure which of adding a new variable or tacking on a fourth parameter to `code:fndef` is less icky (or if there's a better option I'm not thinking of).

ScrelliCopter replied:
I think this is idea is fine because as far as I'm aware nothing else can appear in this spot anyway, maybe the user should still type out the throws because there is a difference between
```
re2c:YYFN = ["lex;Token;throws"];
```
and
```
re2c:YYFN = ["lex;Token;throws(ParseError)"];
```

ScrelliCopter replied:
You still use `try` for calling both but constrained throws need to be wrapped in parentheses. Even if it was possible I think that `["lex;Token;"];` for the unconstrained case is particularly unreadable just to be able to do `["lex;Token;ParseError"];`. I will need to check if `throw(any Error)` is equal to unconstrained `throws` because in that case just assuming the constraint is always in parentheses would be okay because I think `["lex;Token;MyErrorType"];` would be more common than the `["lex;Token;any Error"];` case, I do not know for sure if this is possible tho.

skvadrik replied:
It's fine to have `throw` in `YYFN` def if `any Error` doesn't work out.

The other question is, is it always `"try "` that we want to prepend at call site, or can it be different forms? Do we need the freedom to specify this as well?

skvadrik replied:
There are other options. Exception specifications are popular enough case that we can add something special for them instead of using `YYFN`, like `re2c:fn:throw = "...";` and then have a conditional `.throw` in function calls.

ScrelliCopter replied:
> The other question is, is it always `"try "` that we want to prepend at call site, or can it be different forms? Do we need the freedom to specify this as well?

I don't think there's anything else that fits, there is `try?` and try!` but neither of them make sense in a recursive function context, so I think we should just stick to "try ". The only value I see is being able to swich to `try!` for early prototyping/debugging(?) purposes, so it's really not a factor unless this provides value to other languages.


> There are other options. Exception specifications are popular enough case that we can add something special for them instead of using `YYFN`, like `re2c:fn:throw = "...";` and then have a conditional `.throw` in function calls.

I like this because I think it's clear and users will understand what code it will generate (and thus suddenly getting a bunch of `try` won't be surprising). I can't think off the top of my head other languages with exceptions/psuedo-exception errors that this could provide value to, maybe Java's checked exceptions?

ScrelliCopter replied:
Okay I looked into `throws(any Error)` it is semantically similar to unconstrained throws and appears to work the same in my test program, however there are some significant differences as documented here https://docs.swift.org/swift-book/documentation/the-swift-programming-language/errorhandling/#Specifying-the-Error-Type
The tl;dr is that unbound `throws` allows the compiler the possibility to infer thrown errors and avoid allocating an NSError and that `throws(any Error)` disallows this. If possible we should allow unconstrained throws to be emitted as `throws`. Side note: I did check and `throws()` is invalid, so the syntax file would need to be able to drop the parentheses in this case too.

skvadrik replied:
> If possible we should allow unconstrained throws to be emitted as `throws`.

Sure, thanks for the thorough investigation.

So, to sum up:
- we'll add configuration `re2c:fn:throw` that by default is an empty string, but can be set to a string like `"throw(ParseError)"` or `"throws"`
- syntax configurations `code:fncall`, `code:tailcall` will have a conditional `.throw` that is true if and only if `re2c:fn:throw` is nonempty (for Swift, they will use this conditional to emit additional `"try "`)
- syntax configurations `code:fndecl`, `code:fndef` will also have a conditional `.throw` and a variable `throw` that will expand to the contents of `re2c:fn:throw`

What do you think?

I'm not sure if we should keep `attrs` variable and `.attrs` condition; I used them in C/C++ config to add attributes, and it can be potentially useful, but it's generally not a good idea to add "potentially useful" things. This one does look generic enough to be useful, though... Not sure yet.

skvadrik replied:
Ok, I've made up my mind to revert `attrs`, since it's very easy to add them if needed (and by then we might want to add them in a slightly diffferent form). I'll push the revert soon.

If the plan above looks good to you, do you prefer to work on `re2c:fn:throw` implementation yourself, or would you rather leave it to me? No pressure - just checking in advance so that our work does not collide.

skvadrik replied:
Reverted attributes: https://github.com/skvadrik/re2c/commit/e536d7129fafd66515a8234e7aa54db886391f72.

ScrelliCopter replied:
> If the plan above looks good to you, do you prefer to work on `re2c:fn:throw` implementation yourself, or would you rather leave it to me? No pressure - just checking in advance so that our work does not collide.

I'm completely unfamiliar with the re2c codebase so I probably wouldn't be much help there, if you'd prefer to handle the integration yourself by all means go right ahead. I'm happy to write syntax and tests on the swift side of things if needed.

skvadrik replied:
Sure, I'll take care of it.

skvadrik replied:
Done: https://github.com/skvadrik/re2c/commit/73636c66c52be708601af007f91566cf3f8ec554.

ScrelliCopter replied:
Live: #540

ScrelliCopter replied:
Nice to see that it's useful for other languages, I think this would also be applicable to Java's checked exceptions but I wouldn't know how to work that into any of the tests.

skvadrik commented on test/codegen/swift/__run_all.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
```
You ran this, right? It's easy to forget as this is not hooked in the testing / release scripts.

ScrelliCopter replied:
Yes, I have been using them as the primary way to test.

skvadrik commented on test/codegen/swift/loop_switch/advanced.re:
```diff
@@ -0,0 +1,228 @@
+// re2swift $INPUT -o $OUTPUT -cf
+import Foundation
```
Sorry I didn't notice it before, but we only add codegen models that are not tested in the main examples (non-default ones) here. We can leave these tests if you think there's some additional value, though.

ScrelliCopter replied:
I added all the tests for parity with other backends (and also because I wrote the tests before I wrote the examples). I'm not sure what to do here, should we delete the loop switch tests?

skvadrik replied:
Hmm... yes, let's delete them. All these features are covered by examples.

skvadrik replied:
(Just the loop/switch tests I mean.)

ScrelliCopter replied:
Alrighty, done.

skvadrik commented on test/codegen/swift/nested_fixed_tags.re:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  static func main() {
+    test(call: lexTagsGeneric, string: "\0",   expect: -1)
+    test(call: lexTagsGeneric, string: "ba\0", expect:  1)
+    test(call: lexTagsSimple,  string: "\0",   expect: -1)
+    test(call: lexTagsSimple,  string: "ba\0", expect:  1)
+    test(call: lexCapturesGeneric, string: "\0",  expect: -1)
+    test(call: lexCapturesGeneric, string: "a\0", expect:  0)
+    test(call: lexCapturesSimple,  string: "\0",  expect: -1)
+    test(call: lexCapturesSimple,  string: "a\0", expect:  0)
```
Why the extra terminating nulls here and in other tests in this directory?

ScrelliCopter replied:
I wrote the tests before many of the examples, for them I tend to use `UnsafeBufferPointer<UInt8>` for the lex function because (unlike `UnsafePointer`) it contains bounds info which is automatically checked at runtime in debug. I was using this to guarantee a runtime error if the lexer read past the end of the buffer, but the returned bounds does not contain a null-terminator (and I believe this is the same for [Strings](https://developer.apple.com/documentation/swift/string/withutf8(_:))) so I had to explicitly add them to the literals.

skvadrik commented on include/syntax/swift:
```diff
@@ -0,0 +1,412 @@
+single_quoted_strings = 0;
+indentation_sensitive = 0;
+wrap_blocks_in_braces = 0;
+special_escapes = "\a\b\t\n\v\f\r\"\\";
```
Are these all really supported in Swift? As far as I read here https://docs.swift.org/swift-book/documentation/the-swift-programming-language/stringsandcharacters you don't need `\a, \b, \f, \v` (`\0` can be ignored):

> The escaped special characters \0 (null character), \\ (backslash), \t (horizontal tab), \n (line feed), \r (carriage return), \" (double quotation mark) and \' (single quotation mark)

I guess it doesn't show up on the test because you are using hex literals anyway.

ScrelliCopter replied:
They also don't show up in the [document](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/lexicalstructure/#String-Literals) I was referencing too, I think I just misunderstood the purpose of this configuration (looking at the test I definitely recognised that they didn't exist in the language), I also missed the single quote, probably because I didn't think it would ever need escaping since they're unused.

I've committed something that should hopefully fix this.

skvadrik replied:
The purpose of this configuration is for re2c to know if certain characters need to be rendered in a special way. When it generates the matching clauses, it emits characters like `'a'` or `\xff`, but special escape symbols need custom rendering rules. For swift this is irrelevant at the moment, since characters are rendered as hex numbers.

skvadrik reviewed (changes requested):
Thanks again for all your work. There's quite a few comments but they should all be easy to fix, I'm just nitpicking.

ScrelliCopter commented on examples/swift/01_basic.re:
```diff
@@ -0,0 +1,14 @@
+  */
+}
+
+let string: StaticString = "1234"
+assert(lex(string.utf8Start))
```
Since this was one of the earliest tests/examples I wrote, I was still using StaticStrings everywhere (I don't think you can define literals for them inline like some languages let you, StaticString is also special and has & lacks some features of the typical String type), since the lex function here takes an `UnsafePointer` using a direct string literal with no variable is fine and a little more readable.

ScrelliCopter commented on examples/swift/eof/02_bounds_checking.re:
```diff
@@ -0,0 +1,44 @@
+  let length = str.count
+
+  // Make a copy of the string with YYMAXFILL zeroes at the end.
+  let buffer = ContiguousArray<UInt8>(str) + repeatElement(0, count: yymaxfill)
```
Done.

ScrelliCopter commented on examples/swift/eof/04_fake_sentinel.re:
```diff
@@ -0,0 +1,34 @@
+// re2swift $INPUT -o $OUTPUT --api generic
```
Done.

ScrelliCopter commented on examples/swift/eof/02_bounds_checking.re:
```diff
@@ -0,0 +1,44 @@
+  }
+}
+
+func test(_ str: StaticString, _ expect: Int?) {
+  str.withUTF8Buffer {
+    assert(lex($0) == expect)
+  }
+}
```
In this particular example I'm using `UnsafeBufferPointer<UInt8>` because it contains bounds information (and also doesn't contain a null terminator). Most of the other examples want a null terminator so it's easier to pass the string directly into `UnsafePointer<UInt8>` as Swift will automatically convert it for you (this is primarily to make working with C APIs easier).

If I get the opportunity to type a string I usually type it StaticString as being knowable at compile time has some nice properties if you don't need anything dynamic like string concatenation, and is basically perfect for most of the tests where they are just being treated as read-only contiguous runs of null-terminated UTF-8 bytes. In this particular case getting a buffer pointer out of String would lead to pretty much the same code with a [slightly different API](https://developer.apple.com/documentation/swift/string/withutf8(_:)) (because Swift reasons).

In this case though it might be better to eliminate the test function altogether and just pass the String/StaticString to lex directly.

skvadrik replied:
Got it, thanks!

ScrelliCopter commented on examples/swift/eof/05_fake_sentinel_eof_rule.re:
```diff
@@ -0,0 +1,40 @@
+// re2swift $INPUT -o $OUTPUT --api generic
```
Done.

ScrelliCopter commented on examples/swift/eof/05_fake_sentinel_eof_rule.re:
```diff
@@ -0,0 +1,40 @@
+func lex(_ str: Data) -> Int? {
+  let length = str.count
+
+  var cursor = 0, limit = length, marker = 0
```
Nope, I think that was a leftover from the structure of the previous eof tests.

ScrelliCopter replied:
Assuming you were waiting on this so I went ahead and inlined them.

skvadrik replied:
Thanks!

skvadrik reviewed (approved):
Ready to merge as soon as the check pass. Thank you for addressing all the comments. :)

skvadrik commented:
> Thanks a lot to you for your help. Would you like me to squash the review changes after the checks?

Yes, please do.
